### DISCUSSION TITLE
Improving the timelineVariable API

### DISCUSSION THREAD

**Discussion by jodeleeuw** [jspsych/jsPsych#788] [Q&A] [answered]
Currently there's a hidden parameter in `jsPsych.timelineVariable()` that must be set to true when it is used inside a function. The second parameter controls whether the return value of `timelineVariable()` is a function or the value of the timeline variable directly.

This isn't intuitive, and it's easy to miss in [the documentation](https://www.jspsych.org/core_library/jspsych-core/#jspsychtimelinevariable). It'd be great to find an alternative specification for the function that either made it work in both contexts, or made the API more transparent. 

Any thoughts about what would be a logical redesign here?

**Comment by hauselin** [comment]
Yea, I initially missed that too. Two thoughts/suggestions that don't require too much redesign:

1. Explain this parameter your simple reaction time task demo.
2. Change the parameter name from `call_immediate` to something else like `return_value` (I find `call_immediate` a bit unintuitive here).

**Comment by vijaymarupudi** [comment, marked as answer]
I have an idea for a way which would make timelineVariable(variable_name) to work in both contexts without the second parameter.

* Add a boolean flag variable (`jsPsych.internal.call_immediate`?)
* Use that namespaced flag instead of the second call_immediate parameter of the function.
* It will be false when the user is building the experiment, so it will work as it does without the second argument when the timeline array is being built.
* At execution time, just before [evaluating the functions] in the plugin options, set the namespaced flag to `true`. So when the function runs, the value returned from the `timelineVariable()` function will be the value of interest.
* Immediately after the `evaluateFunctionParameters()` function, set the flag to `false`.

I think this way, backwards compatibility will be maintained (second parameter will be ignored as it's not necessary, jsPsych already has that information) while enabling a more ergonomic API.

[evaluating the functions]: https://github.com/jspsych/jsPsych/blob/83980085ef604c815f0d97ab55c816219e969b84/jspsych.js#L901

**Reply by jodeleeuw** [reply]
I was thinking along these lines too, but I hadn't thought about the way to maintain backwards compatibility or switching the context back and forth, so this is a substantial improvement in design.

I'm trying to think if there are any situations where this would produce an unintended consequence, but so far can't come up with any...

**Reply by vijaymarupudi** [reply]
I believe this is how the React library implements hooks, it thought it was clever when I read about it. Involves some 'global' state, but since Javascript is single threaded, it works elegantly I think!

**Reply by jodeleeuw** [reply]
Bumping this to an issue #883
